I'm not a native speaker so I'm going to waste a lot of time in future to translate my notes

# Анализ комментариев на сайте Hacker News

На сайте Hacker News представлены различные публикации, связанный с IT сферой и не только. 
Мы проанализируем, к каким постам и в какое время пользователи оставляли комментарии.
Для начала просто откроем архив данных и конвертируем его в список списков.

In [1]:
from csv import reader
hn = list(reader(open('hacker_news.csv')))

# Введение
Далее мы просто проанализируем, какие строки и столбцы мы имеем.

In [2]:
print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


# Заголовки и основная часть данных
Для нашего же удобста в анализе данных мы разделим заголовоки и нужные нам данные. 

In [3]:
headers = hn[0]
hn = hn[1:]
print(headers, hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'] [['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


# Разделение постов на три класса
Классы:

    1. "Ask Posts" - посты с вопросами от пользователей. 
    Пример: "Can anyone suggest a good RSS newsreader with a set of tech news feeds?".
    
    2. "Show posts" - публикации с проектами пользователей, публикации о новый продуках или просто что-то интересное.
    Пример: "Space Shooter Retro Game Recreated Using Python".
    
    3. "Other Posts" - этот класс включает в себя всё остальное, этот класс нас не особо интересует так, как в нашем проекте мы будем рассматривать два других класса/категории постов
    Пример: "Hedge Fund Star: We Are Under Assault".

In [4]:
ask_posts, show_posts, other_posts = [],[],[]
for row in hn:
    title = row[1].lower()
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

print('The number of ask posts:', len(ask_posts))
print('The number of show posts:', len(show_posts))
print('The number of other posts:', len(other_posts))

The number of ask posts: 1744
The number of show posts: 1162
The number of other posts: 17194


Для Вашего же удобства ниже мы написали удобный для просмотра код, который наглядно покажет вам, какие темы постов встречаются в Hacker News.

In [5]:
import random as rd
print("Example of ask post:", ask_posts[rd.randint(0, len(ask_posts))][1])
print("Example of show post:", show_posts[rd.randint(0, len(show_posts))][1])
print("Example of other post:", other_posts[rd.randint(0, len(other_posts))][1])

Example of ask post: Ask HN: How to raise a hacker?
Example of show post: Show HN: Demo of live container migration using Virtualbox, Vagrant and ShutIt
Example of other post: Brexit: Wave of racial abuse follows


# Количество комментариев
Теперь нам нужно подсчитать среднее количество комментариев под постами каждого класса/категории. Чтобы было проще находить среднее значение, мы создадим ниже функцию, которая получает на вход список спиков и нужный столбец для подсчёта среднего значения. 

In [6]:
def avg_count(list0, column):
    total_count = 0 
    for row in list0: 
        num = int(row[column])
        total_count += num
    return total_count / len(list0)
        

Ниже мы применим нашу новую функцию к нашим данным, разделённым на два класса.

In [7]:
avg_ask_comments = avg_count(ask_posts, 4)
avg_show_comments = avg_count(show_posts, 4)
print("Average number of ask comments:", avg_ask_comments)
print("Average number of show comments:", avg_show_comments)

Average number of ask comments: 14.038417431192661
Average number of show comments: 10.31669535283993


# Вывод
Мы можем сделать предварительный вывод, что публикации с вопросами вызывают больше ответных действий со стороны пользователей. Это можно обосновать тем, что посты с вопросами нуждаются в отклике аудитории, в ответе, а публикации, рассказывающие о новинках, проектах и так далее в этом не нуждаются. 

Поскольку публикации с вопросами вызывают больший ажиотаж, мы сосредоточимся именно на них. 

# Нахождение количества публикаций и общего количества комментариев в каждом часе суток
Ниже мы создадим два словаря, которые будут состоять из информации о количестве публикаций в час и количестве комментариев в час.

In [8]:
import datetime as dt
result_list = []
counts_by_hour = {}
comments_by_hour = {}

for row in ask_posts:
    result_list.append([row[6], int(row[4])])

for row in result_list:
    date = row[0]
    count_com = row[1]
    hour = dt.datetime.strptime(date, "%m/%d/%Y %H:%M").strftime("%H")
    if hour in counts_by_hour:
        counts_by_hour[hour] += 1 
        comments_by_hour[hour] += count_com
    else:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = count_com

# Поиск среднего количества комментариев в каждом часе суток
Теперь нам нужно поделить количество комментариев в часе на количество постов в этом же часе.

In [9]:
avg_by_hour = []
for hour, count in comments_by_hour.items():
    avg_by_hour.append([hour, count/counts_by_hour[hour]])
avg_by_hour = sorted(avg_by_hour)
print(avg_by_hour)

[['00', 8.127272727272727], ['01', 11.383333333333333], ['02', 23.810344827586206], ['03', 7.796296296296297], ['04', 7.170212765957447], ['05', 10.08695652173913], ['06', 9.022727272727273], ['07', 7.852941176470588], ['08', 10.25], ['09', 5.5777777777777775], ['10', 13.440677966101696], ['11', 11.051724137931034], ['12', 9.41095890410959], ['13', 14.741176470588234], ['14', 13.233644859813085], ['15', 38.5948275862069], ['16', 16.796296296296298], ['17', 11.46], ['18', 13.20183486238532], ['19', 10.8], ['20', 21.525], ['21', 16.009174311926607], ['22', 6.746478873239437], ['23', 7.985294117647059]]


# Сортировка данных от наибольшего к наименьшему
Ниже мы просто сортируем наши данные по убыванию и выводим их в удобном для чтения формате.

In [10]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
sorted_swap = sorted(swap_avg_by_hour, reverse = True)
print("Top 5 Hours for Ask Posts Comments:")
template = "{}: {:.2f} average comments per post"
for c, h in sorted_swap[:5]:
    print(template.format(dt.datetime.strptime(h, "%H").strftime("%H:%M"), c))

Top 5 Hours for Ask Posts Comments:
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


# Вывод
В конце концов, исходя из полученных нами данных, мы можем сделать вывод, что в среднем посты с вопросами, опубликованные в промежуток с 15:00 до 16:00, получают наибольшее количество комментариев - 38.59. Разница между часами с наибольшим и вторым по величине средним числом комментариев состаляет 60%. Важно заметить, что речь идет о Восточно-Американском часовом поясе (UTC-5). Также нельзя не отметить, что в нашем анализе мы учитывали только публикации, под которыми был оставлен хотя бы один комментарий, то есть посты без комментариев не учитывались. Исходя из этого, стоит поправить наши слова, сказав, что среди постов с комментариями наибольшее среднее количество комментариев имеют публикации с вопросами, опубликованные в промежуток между 15:00 и 16:00.

# Дополнительное задание


Создадим функцию "avg_count_per_hour", которая на вход получает основной список, индекс столбца с нужным нам для анализа значением, индекс столбца с датой публикации поста, название класса/категории данных, названия столбца со значениями, значение True или False, позволяющее вывести отсортированные данные на экран, и также значение True или False относительно вывода среднего значения в каждом посту, а на выход функция даёт нам список списков со среднеми значениями и соответствующими часами.

In [12]:
def avg_count_per_hour(input_list, c_index, t_index, name_of_column="", name_of_class="", print_avg_per_hour=False, print_avg_count=False):
    dict_of_post = {}
    dict_of_counts = {}
    list_of_counts = []
    for row in input_list:
        list_of_counts.append([row[t_index], int(row[c_index])])
    for h, c in list_of_counts:
        h = dt.datetime.strptime(h, "%m/%d/%Y %H:%M").strftime("%H")
        if h in dict_of_post:
            dict_of_post[h] += 1
            dict_of_counts[h] += c
        else:
            dict_of_post[h] = 1 
            dict_of_counts[h] = c
    
    avg_list_of_counts = []
    for h, c in dict_of_counts.items():
        avg_list_of_counts.append([c/dict_of_post[h], h])
    avg_list_of_counts = sorted(avg_list_of_counts, reverse=True)
    if print_avg_count:
        print("The average number of", name_of_column, "per post:")
        avg = 0 
        for h, c in list_of_counts:
            avg += c
        print(avg/len(input_list))
    if print_avg_per_hour:
        print("Top 5 for", name_of_class, name_of_column + ":")
        for c, h in avg_list_of_counts[:5]:
            template = "{}: {:.2f} average {} per hour"
            print(template.format(dt.datetime.strptime(h, "%H").strftime("%H:%M"), c, name_of_column))
    return avg_list_of_counts

In [13]:
avg_count_per_hour(ask_posts, 3, 6, "points", "Ask Posts", True, True)
avg_count_per_hour(ask_posts, 4, 6, "comments", "Ask Posts", True, True)
print("")
avg_count_per_hour(show_posts, 3, 6, "points", "Show Posts", True, True)
avg_count_per_hour(show_posts, 4, 6, "comments", "Show Posts", True, True)
print("")
avg_count_per_hour(other_posts, 3, 6, "points", "Other Posts", True, True)
avg_count_per_hour(other_posts, 4, 6, "comments", "Other Posts", True, True)
pass

The average number of points per post:
15.061926605504587
Top 5 for Ask Posts points:
15:00: 29.99 average points per hour
13:00: 24.26 average points per hour
16:00: 23.35 average points per hour
17:00: 19.41 average points per hour
10:00: 18.68 average points per hour
The average number of comments per post:
14.038417431192661
Top 5 for Ask Posts comments:
15:00: 38.59 average comments per hour
02:00: 23.81 average comments per hour
20:00: 21.52 average comments per hour
16:00: 16.80 average comments per hour
21:00: 16.01 average comments per hour

The average number of points per post:
27.555077452667813
Top 5 for Show Posts points:
23:00: 42.39 average points per hour
12:00: 41.69 average points per hour
22:00: 40.35 average points per hour
00:00: 37.84 average points per hour
18:00: 36.31 average points per hour
The average number of comments per post:
10.31669535283993
Top 5 for Show Posts comments:
18:00: 15.77 average comments per hour
00:00: 15.71 average comments per hour
14:

# Выводы:
## Относительно публикаций с вопросами (Ask) 
Можно заметить, что пользователи активнее всего в постах с вопросами в промежуток между 15:00 и 16:00, как в отношении количества комментариев так и в отношении положительных отзывов. Но нужно обозначить, что разница в усреднённом количестве лайков в промежутки 15:00 - 16:00 и 13:00 - 14:00 небольшая, примерно 5.73, но отрезок времени с 13:00 до 14:00 даже не входит в пятёрку наибольших по количеству комментариев промежутков. Таким образом, зависимость между средним количеством комментариев, а также положительных отзывов в одном и том же часе пропадает. Но при этом, промежуток с 16:00 до 17:00 стоит на третьем месте списка количества положительных отзывов в час, а в списке количества комментариев в час на четвертом. Хотя здесь тоже прослеживается зависимость так, как у людей есть больше свободного времени под вечер для написания комментария, ведь это обычно занимает намного больше времени, чем оставления положительного отзыва (лайка), и при этом почти ни что не мещает оставить лайк в обеденный перерыв (15:00). Также напрашивается вывод относительного среднего количества комментариев и лайков по одним постом - разница между этими значениями небольшая, возможно от того, что большинство людей, написавших комментарий поставят лайк, к тому же скорее всего человек, встретившийся с подобной проблемой, напишет об этом в комментариях и поставит лайк. 

## Относительно информационных публикаций (Show) 
Количество лайков значительно больше, чем количество комментвиев, это говорит о том, что посты типа "Show" не нуждаются в комментариях, но при этом количество лайков больше, чем на публикациях с вопросами, это можно обосновать тем, что информационные посты рассчитаны на большую аудиторию.

## Относительно остальных постов (Other)
В постах класса "Other" также отлеживается тенденция разницы между количеством комментариев и лайков, опять же из-за того, что такие публикации не требуют комментариев, нежели посты типа "Ask", которые только и рассчитаны на коммуникацию пользователей. Количество лайков и комментариев в таких постах значительно больше, чем в публикациях "Ask" и "Show", это может означать, что они предназначены для охвата большей аудитории. 

# Окончательный вывод 
Публикации на общие темы вызывают большую реакцию со стороны пользователей.